Xamarin.Forms.Core/VisualStateManager.cs

```diff
@@ -0,0 +1,357 @@
+		readonly IList<VisualStateGroup> _internalList;
+
+		void Validate(IList<VisualStateGroup> groups)
+		{ 
+			// If we have 1 group, no need to worry about duplicate group names
+			if (groups.Count > 1)
+			{
+				if (groups.GroupBy(vsg => vsg.Name).Any(g => g.Count() > 1))
+				{
+					throw new InvalidOperationException("VisualStateGroup Names must be unique");
+				}
+			}
+
+			// State names must be unique within this group list, so pull in all 
+			// the states in all the groups, group them by name, and see if we have
+			// and duplicates
+			if (groups.SelectMany(group => group.States)
+				.GroupBy(state => state.Name)
+				.Any(g => g.Count() > 1))
+			{
+				throw new InvalidOperationException("VisualState Names must be unique");
+			}
+		}
+
+		public VisualStateGroupList() 
+		{
+			_internalList = new WatchAddList<VisualStateGroup>(Validate);
+		}
+
+		void ValidateOnStatesChanged(object sender, EventArgs eventArgs)
+		{
+			Validate(_internalList);
+		}
+
+		public IEnumerator<VisualStateGroup> GetEnumerator()
+		{
+			return _internalList.GetEnumerator();
+		}
+
+		IEnumerator IEnumerable.GetEnumerator()
+		{
+			return ((IEnumerable)_internalList).GetEnumerator();
+		}
+
+		public void Add(VisualStateGroup item)
+		{
+			_internalList.Add(item);
+			item.StatesChanged += ValidateOnStatesChanged;
+		}
+
+		public void Clear()
+		{
+			foreach (var group in _internalList)
+			{
+				group.StatesChanged -= ValidateOnStatesChanged;
+			}
+
+			_internalList.Clear();
+		}
+
+		public bool Contains(VisualStateGroup item)
+		{
+			return _internalList.Contains(item);
+		}
+
+		public void CopyTo(VisualStateGroup[] array, int arrayIndex)
+		{
+			_internalList.CopyTo(array, arrayIndex);
+		}
+
+		public bool Remove(VisualStateGroup item)
+		{
+			item.StatesChanged -= ValidateOnStatesChanged;
+			return _internalList.Remove(item);
+		}
+
+		public int Count => _internalList.Count;
+
+		public bool IsReadOnly => false;
+
+		public int IndexOf(VisualStateGroup item)
+		{
+			return _internalList.IndexOf(item);
+		}
+
+		public void Insert(int index, VisualStateGroup item)
+		{
+			item.StatesChanged += ValidateOnStatesChanged;
+			_internalList.Insert(index, item);
+		}
+
+		public void RemoveAt(int index)
+		{
+			_internalList[index].StatesChanged -= ValidateOnStatesChanged;
+			_internalList.RemoveAt(index);
+		}
+
+		public VisualStateGroup this[int index]
+		{
+			get => _internalList[index];
+			set => _internalList[index] = value;
+		}
+	}
+
+	[RuntimeNameProperty(nameof(Name))]
+	[ContentProperty(nameof(States))]
+	public sealed class VisualStateGroup 
+	{
+		public VisualStateGroup()
+		{
+			States = new WatchAddList<VisualState>(OnStatesChanged);
+		}
+
+		public Type TargetType { get; set; }
+		public string Name { get; set; }
+		public IList<VisualState> States { get; }
+		public VisualState CurrentState { get; internal set; }
+
+		internal VisualState GetState(string name)
+		{
+			foreach (VisualState state in States)
```

Comment:
If this method is called often, does it make sense to switch to for loop? Or is this premature optimisation?

Reply:
I didn't think there'd be a significant difference, since the underlying implementation is using List<T>. But just to be certain I wrote up some quick-and-dirty benchmarking tests to see if `for` would make any difference vs `foreach` here. 

For the first batch, I created VisualStateGroupLists with 1 group, and the group had 2 states (which I think will be a pretty common scenario). I then had the test run through 100,000 iterations of moving randomly between the states (using the `GoToState` method). Between the two implementations (`foreach` and `for`), the test runs took the same amount of time on average.

For the second batch, I created an uglier scenario with VisualStateGroupLists with 10 groups, each group having 10 states. Running that over 100,000 iterations, the `foreach` actually ended up being slightly faster. 

Looking at the tests in the profiler, the bulk of the time spent in the method (in either implementation) is the string comparison. The looping method doesn't end up makeing much difference. 

So for the time being I think we'll stick with `foreach` for clarity. If we start running into performance problems with this later, we can revisit it.

Reply:
The difference isn't in performance but in allocation. foreach generates an iterator which adds some GC pressure. It probably doesn't matter much though.

Reply:
The enumerator for List<T> is a struct, so it's pretty low-impact.